nilearn/_utils/niimg_conversions.py

```diff
@@ -161,176 +244,74 @@ def concat_niimgs(niimgs, dtype=np.float32, accept_4d=False,
-                       niimg.shape))
-            yield niimg
-        except TypeError as exc:
-            exc.args = (('Error encountered while loading image #%d' % i,) + 
-                        exc.args)
-            raise
+    # Case 2: all niimgs are filepaths or memory mapped
+    # -------------------------------------------------
 
+    # Same as above, we do a first pass. It is less costly because we will
+    # only read file headers.
 
-def check_niimg_4d(niimgs, return_iterator=False):
-    """Check that niimg is a proper 4D niimg-like object and load it.
+    # Case 3: niimgs is a generator
+    # -----------------------------
 
-    Parameters
-    ----------
-    niimgs: 4D Niimg-like object
-        See http://nilearn.github.io/building_blocks/manipulating_mr_images.html#niimg.
-        If niimgs is an iterable, checks if data is really 4D. Then,
-        considering that it is a list of niimg and load them one by one.
-        If niimg is a string, consider it as a path to Nifti image and
-        call nibabel.load on it. If it is an object, check if get_data
-        and get_affine methods are present, raise an Exception otherwise.
+    # This is the only case in which we can't browse it several times: we
+    # accumulate data in a list and concatenate it in the end
 
-    return_iterator: boolean
-        If True, an iterator of 3D images is returned. This reduces the memory
-        usage when `niimgs` contains 3D images.
-        If False, a single 4D image is returned. When `niimgs` contains 3D
-        images they are concatenated together.
-
-    Returns
-    -------
-    niimg: 4D nibabel.Nifti1Image or iterator of 3D nibabel.Nifti1Image
-
-    Notes
-    -----
-    This function is the equivalent to check_niimg_3d() for Niimg-like objects
-    with a session level.
-
-    Its application is idempotent.
-    """
-    # Two types of input:
-    # * 4D nifti image
-    # * list of 3D niimgs
-
-    # We get rid of the 4D nifti image case, as it is the simplest case
-    # Use hasattr() instead of isinstance to workaround a Python 2.6/2.7 bug
-    # See http://bugs.python.org/issue7624
-    if isinstance(niimgs, _basestring) or not hasattr(niimgs, "__iter__"):
-        niimgs = load_niimg(niimgs)
-        shape = niimgs.shape
-        if len(shape) != 4:
-            raise TypeError(
-                "Data must be a 4D Niimg-like object but you provided an "
-                "image of shape %s. See "
-                "http://nilearn.github.io/building_blocks/"
-                "manipulating_mr_images.html#niimg." % (shape, ))
-        if return_iterator:
-            return (_index_niimgs(niimgs, i) for i in range(shape[3]))
-        else:
-            return niimgs
-
-    # We now have 3 types of input:
-    # * a true list
-    # * an iterator
-    # * a generator
+    target_fov = 'first' if auto_resample else None
+    first_niimg = None
+    if not inspect.isgenerator(niimgs):
```

Comment:
```python
import itertools

import numpy as np
from nibabel import Nifti1Image

from nilearn import _utils

affine = np.eye(4)
img_3d = Nifti1Image(np.ones((10, 10, 10)), affine)

input_iterator = itertools.chain([img_3d], [img_3d])
img_4d = _utils.check_niimg_4d(input_iterator)
print(img_4d.shape)  # (10, 10, 10, 2) which is correct
print(img_4d.get_data())  # uninitialised data rather than ones
```

This happens because the first loop to compute the shape works fine but the second loop to set the data never gets run because the iterator has already been consumed in the first one.

All I am trying to say is that the check should be whether `niimgs` is an iterator rather than a generator.